File: frontend/src/app/features/feed/feed.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface Post {
  id: string;
  content: string;
  imageUrl?: string;
  authorName: string;
  authorUsername: string;
  authorId: string;
  authorProfilePictureUrl?: string; // Add this
  createdAt: string;
  likesCount: number;
  commentCount: number;
  likedByCurrentUser: boolean;
  plantId?: string;
  plantNickname?: string;
  plantTag?: string;
  originalPost?: Post; // Recursively defined
  repostCount: number;
  isRepostedByCurrentUser: boolean;
}

export interface CommentRequest {
  content: string;
}

@Injectable({
  providedIn: 'root'
})
export class FeedService {
  private http = inject(HttpClient);
  private apiUrl = 'http://localhost:8080/api/v1/feed';

  getFeed(page: number = 0, size: number = 10, plant?: string, query?: string): Observable<{ content: Post[] }> {
    let url = `${this.apiUrl}?page=${page}&size=${size}`;
    if (plant) {
      url += `&plant=${encodeURIComponent(plant)}`;
    }
    if (query) {
      url += `&q=${encodeURIComponent(query)}`;
    }
    return this.http.get<{ content: Post[] }>(url);
  }

  getPostById(id: string): Observable<Post> {
    // Note: Backend endpoint is /api/v1/posts/{id}, which is distinct from /api/v1/feed
    // Construct the URL directly or update apiUrl base.
    // For now, I'll hardcode the replacement since apiUrl is .../feed
    const baseUrl = this.apiUrl.replace('/feed', '/posts');
    return this.http.get<Post>(`${baseUrl}/${id}`);
  }

  createPost(caption: string, file?: File, plantId?: string, plantTag?: string): Observable<Post> {
    const formData = new FormData();
    formData.append('caption', caption);
    if (file) {
      formData.append('file', file);
    }
    if (plantId) {
      formData.append('plantId', plantId);
    }
    if (plantTag) {
      formData.append('plantTag', plantTag);
    }
    return this.http.post<Post>(this.apiUrl, formData);
  }

  editPost(postId: string, caption: string, plantTag?: string | null): Observable<Post> {
    const formData = new FormData();
    formData.append('caption', caption);
    if (plantTag) {
      formData.append('plantTag', plantTag);
    }
    return this.http.put<Post>(`${this.apiUrl}/${postId}`, formData);
  }

  deletePost(postId: string): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${postId}`);
  }

  likePost(postId: string): Observable<void> {
    return this.http.post<void>(`${this.apiUrl}/${postId}/like`, {});
  }

  repostPost(postId: string): Observable<void> {
    return this.http.post<void>(`${this.apiUrl}/${postId}/repost`, {});
  }

  commentOnPost(postId: string, request: CommentRequest): Observable<void> {
    return this.http.post<void>(`${this.apiUrl}/${postId}/comment`, request);
  }
}

File: backend/src/main/java/com/plantsocial/backend/service/FeedService.java
package com.plantsocial.backend.service;

import com.plantsocial.backend.dto.PostResponse;
import com.plantsocial.backend.model.Plant;
import com.plantsocial.backend.model.Post;
import com.plantsocial.backend.model.PostLike;
import com.plantsocial.backend.notification.model.NotificationType;
import com.plantsocial.backend.notification.NotificationService;
import com.plantsocial.backend.user.User;
import com.plantsocial.backend.repository.CommentRepository;
import com.plantsocial.backend.repository.PlantRepository;
import com.plantsocial.backend.repository.PostLikeRepository;
import com.plantsocial.backend.repository.PostRepository;
import com.plantsocial.backend.user.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.multipart.MultipartFile;

import java.util.UUID;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

@Service
@RequiredArgsConstructor
public class FeedService {

    private static final Logger log = LoggerFactory.getLogger(FeedService.class);

    private final PostRepository postRepository;
    private final PostLikeRepository postLikeRepository;
    private final CommentRepository commentRepository;
    private final PlantRepository plantRepository;
    private final UserRepository userRepository;
    private final FileStorageService fileStorageService;
    private final NotificationService notificationService;

    public Page<PostResponse> getFeed(Pageable pageable, String plant, String query) {
        System.out.println("DEBUG: Entering getFeed");
        User currentUser = getCurrentUser();
        System.out.println("DEBUG: Current user resolved: " + (currentUser != null ? currentUser.getEmail() : "null"));

        Page<Post> posts;
        if (query != null && !query.isBlank()) {
            System.out.println("DEBUG: Searching posts by query: " + query);
            posts = postRepository.findByContentContainingIgnoreCaseOrderByCreatedAtDesc(query.trim(), pageable);
        } else if (plant != null && !plant.isBlank()) {
            System.out.println("DEBUG: Fetching posts by plant tag: " + plant);
            posts = postRepository.findByPlantTagIgnoreCaseOrderByCreatedAtDesc(plant.trim(), pageable);
        } else {
            System.out.println("DEBUG: Fetching all posts");
            posts = postRepository.findAllByOrderByCreatedAtDesc(pageable);
        }
        System.out.println("DEBUG: Posts fetched: " + posts.getTotalElements());

        return posts.map(post -> {
            try {
                return mapToPostResponse(post, currentUser);
            } catch (Exception e) {
                System.out.println("DEBUG: Error mapping post " + post.getId());
                e.printStackTrace();
                throw e;
            }
        });
    }

    public PostResponse getPostById(UUID postId) {
        User currentUser = getCurrentUser();
        Post post = postRepository.findById(postId)
                .orElseThrow(() -> new IllegalArgumentException("Post not found"));
        return mapToPostResponse(post, currentUser);
    }

    @Transactional
    public PostResponse createPost(String caption, MultipartFile file, UUID plantId, String plantTag) {
        User user = getCurrentUser();

        String imageUrl = null;
        if (file != null && !file.isEmpty()) {
            imageUrl = fileStorageService.storeFile(file);
        }

        Plant plant = null;
        if (plantId != null) {
            plant = plantRepository.findById(plantId).orElse(null);
        }

        Post post = Post.builder()
                .content(caption)
                .imageUrl(imageUrl)
                .author(user)
                .plant(plant)
                .plantTag(plantTag != null && !plantTag.isBlank() ? plantTag.trim() : null)
                .build();
        Post savedPost = postRepository.save(post);
        return mapToPostResponse(savedPost, user);
    }

    @Transactional
    public PostResponse editPost(UUID postId, String caption, String plantTag) {
        User user = getCurrentUser();
        Post post = postRepository.findById(postId)
                .orElseThrow(() -> new IllegalArgumentException("Post not found"));
        post.setContent(caption);
        // Set or clear plantTag
        post.setPlantTag(plantTag != null && !plantTag.isBlank() ? plantTag.trim() : null);
        Post savedPost = postRepository.save(post);
        return mapToPostResponse(savedPost, user);
    }

    @Transactional
    public void deletePost(UUID postId) {
        Post post = postRepository.findById(postId)
                .orElseThrow(() -> new IllegalArgumentException("Post not found"));
        // CascadeType.ALL on comments/likes handles cleanup automatically
        postRepository.delete(post);
    }

    @Transactional
    public void likePost(UUID postId) {
        User user = getCurrentUser();
        Post post = postRepository.findById(postId)
                .orElseThrow(() -> new IllegalArgumentException("Post not found"));

        if (postLikeRepository.existsByPostAndUser(post, user)) {
            postLikeRepository.deleteByPostAndUser(post, user);
        } else {
            PostLike like = PostLike.builder()
                    .post(post)
                    .user(user)
                    .build();
            postLikeRepository.save(like);

            // Notify post author
            notificationService.createNotification(
                    post.getAuthor(),
                    user,
                    NotificationType.LIKE,
                    user.getFullName() + " liked your post",
                    post.getId());
        }
    }

    private User getCurrentUser() {
        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        if (authentication == null || !authentication.isAuthenticated()
                || "anonymousUser".equals(authentication.getPrincipal())) {
            return null;
        }

        Object principal = authentication.getPrincipal();
        String username;
        if (principal instanceof UserDetails) {
            username = ((UserDetails) principal).getUsername();
        } else {
            username = principal.toString();
        }
        return userRepository.findByEmail(username)
                .orElseThrow(() -> new UsernameNotFoundException("User not found"));
    }

    public PostResponse mapToPostResponsePublic(Post post, User currentUser) {
        return mapToPostResponse(post, currentUser);
    }

    private PostResponse mapToPostResponse(Post post, User currentUser) {
        long likesCount = postLikeRepository.countByPost(post);
        long commentCount = commentRepository.countByPost(post);
        boolean likedByCurrentUser = currentUser != null && postLikeRepository.existsByPostAndUser(post, currentUser);

        // Repost logic
        long repostCount = postRepository.countByRepostOfId(post.getId());
        boolean isRepostedByCurrentUser = currentUser != null
                && postRepository.existsByAuthorIdAndRepostOfId(currentUser.getId(), post.getId());

        UUID plantId = post.getPlant() != null ? post.getPlant().getId() : null;
        String plantNickname = post.getPlant() != null ? post.getPlant().getNickname() : null;

        return new PostResponse(
                post.getId(),
                post.getContent(),
                post.getImageUrl(),
                post.getAuthor().getFullName(),
                post.getAuthor().getHandle(),
                post.getAuthor().getId(),
                post.getCreatedAt(),
                likesCount,
                commentCount,
                likedByCurrentUser,
                plantId,
                plantNickname,
                post.getPlantTag(),
                post.getAuthor().getProfilePictureUrl(),
                post.getRepostOf() != null ? mapToPostResponse(post.getRepostOf(), currentUser) : null,
                repostCount,
                isRepostedByCurrentUser);
    }

    @Transactional
    public void repostPost(UUID postId) {
        User user = getCurrentUser();
        log.info("repostPost called: postId={}, userId={}", postId, user.getId());

        Post targetPost = postRepository.findById(postId)
                .orElseThrow(() -> new IllegalArgumentException("Post not found: " + postId));

        // Flatten reposts: Always repost the ROOT original post
        Post originalPost = targetPost;
        while (originalPost.getRepostOf() != null) {
            originalPost = originalPost.getRepostOf();
        }
        log.info("Original post resolved: originalPostId={}", originalPost.getId());

        // Toggle Repost Logic
        boolean alreadyReposted = postRepository.existsByAuthorIdAndRepostOfId(user.getId(), originalPost.getId());
        log.info("Already reposted: {}", alreadyReposted);

        if (alreadyReposted) {
            // Undo repost
            java.util.Optional<Post> existingRepost = postRepository.findByAuthorIdAndRepostOfId(user.getId(),
                    originalPost.getId());
            existingRepost.ifPresent(repost -> {
                log.info("Deleting repost: repostId={}", repost.getId());
                postRepository.delete(repost);
                postRepository.flush();
            });
        } else {
            // Create new repost
            Post repost = Post.builder()
                    .author(user)
                    .repostOf(originalPost)
                    .build();

            log.info("Saving new repost for originalPostId={}", originalPost.getId());
            postRepository.save(repost);

            // Notify original author (only if not self-repost)
            if (!originalPost.getAuthor().getId().equals(user.getId())) {
                notificationService.createNotification(
                        originalPost.getAuthor(),
                        user,
                        NotificationType.REPOST,
                        user.getFullName() + " reposted your post",
                        originalPost.getId());
            }
        }
    }
}
